lib/spack/docs/module_file_support.rst

```diff
@@ -0,0 +1,664 @@
+   libelf@0.8.13%gcc@4.4.7 arch=linux-debian7-x86_64
+   libelf@0.8.13%intel@15.0.0 arch=linux-debian7-x86_64
+
+You can either type the ``spack load`` command again with a fully
+qualified argument, or you can add just enough extra constraints to
+identify one package.  For example, above, the key differentiator is
+that one ``libelf`` is built with the Intel compiler, while the other
+used ``gcc``.  You could therefore just type:
+
+.. code-block:: console
+
+   $ spack load libelf %intel
+
+To identify just the one built with the Intel compiler.
+
+=======================
+Module files generation
+=======================
+
+Module files are generated by post-install hooks after the successful
+installation of a package. They are placed in the following directories
+under the Spack root:
+
+  +--------------------------------+------------------------------------+
+  |                                |  **Module files root**             |
+  +================================+====================================+
+  |  **Environment Modules**       |  share/spack/modules               |
+  +--------------------------------+------------------------------------+
+  |  **Dotkit**                    |   share/spack/dotkit               |
+  +--------------------------------+------------------------------------+
+  |  **LMod Hierarchical Modules** |   share/spack/lmod                 |
+  +--------------------------------+------------------------------------+
+
+
+Though Spack ships with sensible defaults for the generation of module files,
+one can customize many aspects of it to accommodate package or site specific needs.
+These customizations are enabled by either:
+
+ 1. overriding certain callback APIs in the Python packages
+ 2. writing specific rules in the ``modules.yaml`` configuration file
+
+The former method fits best cases that are site independent, e.g. injecting variables
+from language interpreters into their extensions. The latter instead permits to
+fine tune the content, naming and creation of module files to meet site specific conventions.
+
+--------------------------
+Overriding ``Package`` API
+--------------------------
+
+There are two methods that can be overridden in any ``package.py`` to affect the
+content of generated module files. The first one is:
+
+.. code-block:: python
+
+   def setup_environment(self, spack_env, run_env):
+       """Set up the compile and runtime environments for a package."""
+       pass
+
+and can alter the content of *the same package where it is overridden*
+by adding actions to ``run_env``. The second method is:
+
+.. code-block:: python
+
+   def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+       """Set up the environment of packages that depend on this one"""
+       pass
+
+and has similar effects on module file of dependees. Even in this case
+``run_env`` must be filled with the desired list of environment modifications.
+
```

Comment:
Which parameters is `setup_dependent_environment()` supposed to modify?

I think it would be simpler if we had a single `setup_environment()` that set up all the environments the Spack package wishes to set up.  Why is this split up in two separate methods?